push/ups_integration_with_openshift.md

```diff
@@ -0,0 +1,190 @@
+```
+
+A sidecar will be used to sync these Push Variant credentials from OpenShift to UPS.
+The Variant ID and Config will be set by the sidecar after creation of the Variant in UPS.
+OpenShift will be the source of truth for Variants linked to Mobile Clients.
+The sycning process in the sidecar will need to consider:
+
+* Syncing *From* OpenShift *to* UPS
+  * New Secrets mean a New Variant is created
+  * Updates to Secrets overwrite the Variant
+  * Deleting a Secret will delete the Variant
+  * Deleting a Variant directly in UPS will cause it to be recreated
+* Syncing *From* UPS *to* OpenShift
+  * Variant ID will be annotated on the Secret after creation
+  * Variant Config will be annotated on the Secret and kept up to date (UPS is the source of truth for this config)
+
+Here is an example workflow showing the UI touchpoints:
+
+* developer creates a Cordova App (MobileClient) in OpenShift
+* developer provisions UPS from the Catalog
+* developer sets up the FCM/Android credentials for push for the Cordova App via *new* UI screens in OpenShift
+* this creates a Secret to store those credentials
+* the UPS sidecar syncs these credentials from the Secret to a new Android Variant in UPS
+* the Variant config is retrieved from UPS and stored an annotation on the Secret
+* the generated mobile-services.json file includes the correct Variant config
+
+```json
+{
+  "version": 1,
+  "clusterName": "https://example.com",
+  "namespace": "myproject",
+  "clientId": "app-cordova",
+  "services": [
+    {
+      "id": "ups",
+      "name": "ups",
+      "type": "ups",
+      "url": "https://ups.example.com",
+      "config": {
+        "android": {
+          "variantId": "some-variant-id",
+          "variantSecret": "some-secret",
+          "senderId": "some-sender-id"
+        }
+      }
+    }
+  ]
+}
+```
+
+A follow on flow would be:
+
+* developer sets up iOS credentials for push for the same Codova App via *new* UI screens in OpenShift
+* these credentials as stored in the Secret
+* on next sync (in the sidecar), an iOS variant is created in UPS
+* the Variant config is retrieved from UPS and stored as annotations in the Secret
+* the generated mobile-services.json file now includes both the iOS and Android Variants config
+
+```json
+{
+  "version": 1,
+  "clusterName": "https://example.com",
+  "namespace": "myproject",
+  "clientId": "app-cordova",
+  "services": [
+    {
+      "id": "ups",
+      "name": "ups",
+      "type": "ups",
+      "url": "https://ups.example.com",
+      "config": {
+        "android": {
+          "variantId": "some-variant-id",
+          "variantSecret": "some-secret",
+          "senderId": "some-sender-id"
+        },
+        "ios": {
+          "variantId": "some-variant-id",
+          "variantSecret": "some-secret"
+        }
+      }
+    }
+  ]
+}
+```
+
+The developer is free to choose which Variants, from all available Variants in the Push Appplication in UPS, are linked to each Mobile Client.
```

Comment:
Isn't the linking done when the annotations/labels are added to the secrets associating them with a Mobile Client? Would this mean that a user can later link an existing configuration (and thus it's secret) to a different mobile client?

Reply:
This section needs revision.
My thinking is we could allow the user to see all variants in the OpenShift UI and let them pick & choose which ones are linked.
I think a first pass would be to allow 1 type of each allowed variant for the Mobile Client type.
To switch between e.g. a development and release certificate, the developer can upload to replace the cert in the variant.

A follow up change could allow switching between known variants (using `oc edit` or new UI screens)
This means there could be variants in UPS that aren't linked to a Mobile Client.
This seems OK to me to allow that, but needs more thought on the execution.